### PR TITLE
feat(client): ts-migrate multiple files

### PR DESCRIPTION
**Pull request by vishwasnavadak**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Related to #42256, #43778

Migrated Files

1. /client/src/templates/Introduction/Intro.js 
2. /client/src/templates/Introduction/SuperBlockIntro.js

### CODE REVIEW

**Comment by ahmadabdolsaheb** [comment]
@vishwasnavadak, It the has been an issue with the rebase that is why it has picked up some translation files among other commits. 

I suggest force pushing your changes after a clean rebase, and let us know if you are still experiencing issues.

**Comment by vishwasnavadak** [comment]
> @vishwasnavadak, It the has been an issue with the rebase that is why it has picked up some translation files among other commits.
> 
> I suggest force pushing your changes after a clean rebase, and let us know if you are still experiencing issues.

Thanks for this suggestion. I did that, Now the history seems cleaner without the translation changes. Let me know if there's anything else to in this PR.

File: client/src/components/formHelpers/form-validators.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  return (value: string): string =>
+    // TODO: Fix the types according to the validator type and null condition
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    validators.reduce(
+      (error: string, validator: ((a: string) => void) | null) =>
+        error ?? validator?.(value),
+      null
+    );
+}
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-validators.tsx:38]:
```suggestion
export type Validator = (value: string) => React.ReactElement | null;
export function composeValidators(...validators: (Validator | null)[]) {
  return (value: string): ReturnType<Validator> | null =>
    validators.reduce(
      (error: ReturnType<Validator>, validator) =>
        error ?? (validator ? validator(value) : null),
      null
    );
}

```

File: client/src/components/formHelpers/form-fields.tsx [hunk 2]
```diff
@@ -44,22 +35,30 @@ function FormFields(props) {
...
       try {
         normalizeUrl(value, { stripWWW: false });
       } catch (err) {
-        validationError = err.message;
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        validationError = err.message as string;
       }
     }
-    const validationWarning = composeValidators(
+    const validationWarning: string = composeValidators(
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-fields.tsx:53]:
```suggestion
    const validationWarning = composeValidators(
```

File: client/src/components/formHelpers/form-fields.tsx [hunk 3]
```diff
@@ -44,22 +35,30 @@ function FormFields(props) {
...
+    value: string,
+    error: unknown,
+    isURL: boolean,
+    name: string
+  ) => {
+    let validationError: string | undefined;
     if (value && isURL) {
       try {
         normalizeUrl(value, { stripWWW: false });
       } catch (err) {
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-fields.tsx:48]:
```suggestion
      } catch (err: unkown) {
```

File: client/src/components/formHelpers/form-fields.tsx [hunk 4]
```diff
@@ -44,22 +35,30 @@ function FormFields(props) {
...
-        validationError = err.message;
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        validationError = err.message as string;
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-fields.tsx:50]:
```suggestion
        validationError = (err as { message?: string })?.message;
```

File: client/src/components/formHelpers/form-fields.tsx [hunk 5]
```diff
@@ -95,6 +94,7 @@ function FormFields(props) {
                       required={required.includes(name)}
                       rows={4}
                       type={type}
+                      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-fields.tsx:97]:
```suggestion
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Thank you, for taking the time to contribute. Please see the recommended changes.

Also, at this point, we should not be using any `ts-ignore` or similar comments, unless absolutely necessary. The suggestions I have given start you off on the right track, but the rest of the types still need to be correctly annotated.

Let us know if you need any help.

**Comment by vishwasnavadak** [comment]
Thanks for the change suggestions. I'll try to make the fixes as per your suggestion. I had added some `ts-ignore` comments since I wasn't sure what is the right type to be applied. This helps. 

I'll update the PR with these changes

File: client/src/components/formHelpers/form-fields.tsx [hunk 6]
```diff
@@ -44,13 +35,18 @@ function FormFields(props) {
...
+    isURL: boolean,
+    name: string
+  ) => {
+    let validationError: string | undefined;
     if (value && isURL) {
       try {
         normalizeUrl(value, { stripWWW: false });
-      } catch (err) {
-        validationError = err.message;
+      } catch (err: unkown) {
```
Comment (ojeytonwilliams) [inline client/src/components/formHelpers/form-fields.tsx:48]:
```suggestion
      } catch (err: unknown) {
```

File: client/src/components/formHelpers/index.tsx [hunk 7]
```diff
@@ -1,27 +1,44 @@
...
+    errors: [],
+    invalidValues: []
+  };
   const urlValues = Object.keys(values).reduce((result, key) => {
-    let value = values[key];
-    const nullOrWarning = composeValidators(
+    // eslint-disable-next-line
+    //@ts-ignore
+    let value: string = values[key] as string;
+    const nullOrWarning: string = composeValidators(
```
Comment (ojeytonwilliams) [inline client/src/components/formHelpers/index.tsx:41]: Unfortunately, client/src/templates/Challenges/projects/solution-form.tsx is misleading, the `ValidatedValues` type is incorrect.  To fix this, I would infer the type from `composeValidators` and correct `solution-form.tsx`'s typing to use that.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Thanks for all your hard work on this, @vishwasnavadak. I've added some suggestions about how to fix the remaining type errors.  

Also, a simple way to confirm that everything is as it should be is to run `npm run lint:ts`.

**Comment by vishwasnavadak** [comment]
@ojeytonwilliams thanks for the help. I have done some of those changes in local, hadn't got time to run the tests and update the PR. Will take your suggestions and update this in couple of days.

File: client/src/components/formHelpers/index.tsx (hunk 7 above)
Comment (vishwasnavadak) [reply client/src/components/formHelpers/index.tsx:41]: @ojeytonwilliams I'm not sure what do you mean by referring the validatedValues type from ComposeValidators.

Comment (ojeytonwilliams) [reply client/src/components/formHelpers/index.tsx:41]: So, the `ValidatedValues` type in client/src/templates/Challenges/projects/solution-form.tsx is not correct.  What you can do is start from client/src/components/formHelpers/form-validators.tsx and set the validators return type to `JSX.Element | null`, then see what other changes that requires.

**Comment by vishwasnavadak** [comment]
Quick Update: I was able to update the types to correct types. There's one last thing I'm trying to fix on the conflicted file. Will update that soon.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hi @vishwasnavadak I think we're nearly there.  

Aside from the one problem with `toggleBlock`, the only issue is naming. Types should be CamelCase and not end with the word `Type` since that's redundant.  I've made a few suggestions below:

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Just one thing:

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍 

Although, I am a little confused about the updated package 🤷‍♂️ - problem for Renovate, I guess

**Pull request merged**